polygeist/lib/polygeist/Passes/ConvertPolygeistToLLVM.cpp

```diff
@@ -45,16 +49,513 @@
+                    : increment;
+    }
+    const auto elementPtrType = getTypeConverter()->convertType(type);
+    if (!elementPtrType)
+      return {};
+    return index
+               ? rewriter.create<LLVM::GEPOp>(loc, elementPtrType, base, index)
+               : base;
+  }
+};
+
+struct LoadMemRefOpLowering : public MemAccessLowering {
+  LoadMemRefOpLowering(LLVMTypeConverter &typeConverter,
+                       PatternBenefit benefit = 1)
+      : MemAccessLowering{memref::LoadOp::getOperationName(),
+                          &typeConverter.getContext(), typeConverter, benefit} {
+  }
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> args,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto loadOp = cast<memref::LoadOp>(op);
+    if (!canBeLoweredToBarePtr(loadOp.getMemRefType()))
+      return failure();
+    memref::LoadOp::Adaptor adaptor{args};
+    const auto DataPtr = getStridedElementBarePtr(
+        loadOp.getLoc(), loadOp.getMemRefType(), adaptor.getMemref(),
+        adaptor.getIndices(), rewriter);
+    if (!DataPtr)
+      return failure();
+    rewriter.replaceOpWithNewOp<LLVM::LoadOp>(op, DataPtr);
+    return success();
+  }
+};
+
+struct StoreMemRefOpLowering : public MemAccessLowering {
+  StoreMemRefOpLowering(LLVMTypeConverter &typeConverter,
+                        PatternBenefit benefit = 1)
+      : MemAccessLowering{memref::StoreOp::getOperationName(),
+                          &typeConverter.getContext(), typeConverter, benefit} {
+  }
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> args,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto storeOp = cast<memref::StoreOp>(op);
+    if (!canBeLoweredToBarePtr(storeOp.getMemRefType()))
+      return failure();
+    memref::StoreOp::Adaptor adaptor{args};
+    const auto DataPtr = getStridedElementBarePtr(
+        storeOp.getLoc(), storeOp.getMemRefType(), adaptor.getMemref(),
+        adaptor.getIndices(), rewriter);
+    if (!DataPtr)
+      return failure();
+    rewriter.replaceOpWithNewOp<LLVM::StoreOp>(op, adaptor.getValue(), DataPtr);
+    return success();
+  }
+};
+
+struct SignatureConversionPattern : public ConversionPattern {
+  using ConversionPattern::ConversionPattern;
+
+protected:
+  /// States whether a signature should be converted
+  ///
+  /// A signature should be converted if it contains at least a memref type and
+  /// all of the memref types in it can be lowered to a bare ptr.
+  static bool shouldConvertSignature(TypeRange types) {
+    bool hasMT{false};
+    for (auto type : types) {
+      if (auto mt = type.dyn_cast<MemRefType>()) {
+        if (!canBeLoweredToBarePtr(mt))
+          return false;
+        hasMT = true;
+      }
+    }
+    return hasMT;
+  }
+};
+
+/// See mlir/lib/Dialect/Func/Transforms/FuncConversions.cpp
+///
+/// Copied here to be able to pass custom benefit.
```

Comment:
Curious why do we need to pass a custom benefit. Perhaps explain in the comment.

Reply:
ok